packages/@aws-cdk/aws-lambda/lib/function-url.ts

```diff
@@ -0,0 +1,195 @@
+  /**
+   * The DELETE method deletes the specified resource.
+   */
+  DELETE = 'DELETE',
+}
+
+/**
+ * Specifies a cross-origin access property for a function URL
+ */
+export interface CorsProperty {
+  /**
+   * Whether to allow cookies or other credentials in requests to your function URL.
+   *
+   * @default false
+   */
+  readonly allowCredentials?: boolean;
+
+  /**
+   * Headers that are specified in the Access-Control-Request-Headers header.
+   *
+   * @default - No headers allowed.
+   */
+  readonly allowedHeaders?: string[];
+
+  /**
+   * An HTTP method that you allow the origin to execute.
+   *
+   * @default - No methods allowed.
+   */
+  readonly allowedMethods?: HttpMethods[];
+
+  /**
+   * One or more origins you want customers to be able to access the bucket from.
+   *
+   * @default - No origins allowed.
+   */
+  readonly allowedOrigins?: string[];
+
+  /**
+   * One or more headers in the response that you want customers to be able to access from their applications.
+   *
+   * @default - No headers exposed.
+   */
+  readonly exposedHeaders?: string[];
+
+  /**
+   * The time in seconds that your browser is to cache the preflight response for the specified resource.
+   *
+   * @default - No caching.
+   */
+  readonly maxAge?: number;
+}
+
+/**
+ *
+ */
+export interface IFunctionUrl extends IResource {
+  /**
+   * The url of the Lambda function.
+   *
+   * @attribute FunctionUrl
+   */
+  readonly url: string;
+
+  /**
+   * The ARN of the function this URL refers to
+   *
+   * @attribute FunctionArn
+   */
+  readonly functionArn: string;
+}
+
+/**
+ * Options to add a url to a Lmabda function
+ */
+export interface FunctionUrlOptions {
+  /**
+   * The type of authentication that your function URL uses.
+   */
+  readonly authType: FunctionUrlAuthType;
+
+  /**
+   * The cross-origin resource sharing (CORS) settings for your function URL.
+   *
+   * @default - No CORS configuration.
+   */
+  readonly cors?: CorsProperty;
+}
+
+/**
+ * Properties for a FunctionUrl
+ */
+export interface FunctionUrlProps extends FunctionUrlOptions {
+  /**
+   * The function to which this url refers.
+   */
+  readonly function: IFunction;
+
+  /**
+   * The alias name.
+   *
+   * @default - No alias, added to the unpublished version of the function.
+   */
+  readonly qualifier?: string;
+}
+
+/**
+ * Defines a Lambda function url
+ *
+ * @resource AWS::Lambda::Url
+ */
+export class FunctionUrl extends Resource implements IFunctionUrl {
+  /**
+   * The url of the Lambda function.
+   */
+  public readonly url: string;
+
+  /**
+   * The ARN of the function this URL refers to
+   */
+  public readonly functionArn: string;
+
+  /**
+   * The function to which this url refers.
+   */
+  public readonly function: IFunction;
+
+  constructor(scope: Construct, id: string, props: FunctionUrlProps) {
+    super(scope, id);
+
+    const resource: CfnUrl = new CfnUrl(this, 'Resource', {
+      authType: props.authType,
```

Comment:
if `authType` is `NONE` you still need to add a permission to Lambda resource policy:

```ts
this.addPermission('FunctionUrlPermission', {
  principal: new iam.AnyPrincipal(),
  action: 'lambda:InvokeFunctionUrl',
  resource: this.functionArn,
  functionUrlAuthType: props.authType, // this key needs to be added to `addPermission()`
});
```

Reply:
Aha. Does that mean we also need `lambda:InvokeFunctionUrl` for AWS_IAM style permissions?

Reply:
No, it's not required because the IAM role can have the proper permission.

Reply:
Yes but is there or is there not a distinction between `lambda:InvokeFunction` and `lambda:InvokeFunctionUrl` ?

Reply:
There is a distinction. Only `lambda:InvokeFunctionUrl` allows to invoke via the function url. So if the function url is configured with `AWS_IAM` and an IAM principal wants to invoke it via the url it needs the `lambda:InvokeFunctionUrl` permission or the resource based policy of the function must allow `lambda:InvokeFunctionUrl` for this principal.

There's no automatic grant to implement.

A `fn.grantInvokeUrl()` could be nice for identity based permissions and for resource based permissions users will be able to use `addPermission()` with the proper action.